File: 3/main.py
#!/usr/bin/python3

# 533775
#

import sys


class Number:
    def __init__(self, sidx, yidx):
        self.sidx = sidx
        self.yidx = yidx
        self.value = 0

    def __add__(self, number):
        self.value = self.value * 10 + number
        return self

    def __len__(self):
        return len(str(self.value))

    def __repr__(self):
        return f"({self.value}, yidx: {self.yidx}, sidx: {self.sidx}, len: {len(self)})"


def extract_numbers(line, yidx):
    numbers = []

    number = None

    i = 0
    while i < len(line):
        if line[i].isdigit():
            if number is None:
                number = Number(i, yidx)
                number += int(line[i])
            else:
                number += int(line[i])
        else:
            if number is not None:
                numbers.append(number)
                number = None
        i += 1

    return numbers


lines = map(lambda line: line.strip(), sys.stdin.readlines())
lines = list(lines)

# append dots around whole map
x = '.' * len(lines[0])
lines.insert(0, x)
lines.append(x)

lines = map(lambda line: '.' + line + '.', lines)
lines = list(lines)


def flatten(list):
    return [item for sublist in list for item in sublist]


numbers = [extract_numbers(line, yidx) for yidx, line in enumerate(lines)]
numbers = filter(lambda x: len(x) > 0, numbers)
numbers = flatten(numbers)
numbers = list(numbers)


def print_lines(lines):
    print('====================')
    for line in lines:
        print(line)
    print('====================')


def get_border(number, lines):
    x1 = number.sidx-1
    x2 = number.sidx + len(number)

    border = ''
    border += lines[number.yidx - 1][x1: x2+1]
    border += lines[number.yidx + 1][x1: x2+1]
    border += lines[number.yidx][x1]
    border += lines[number.yidx][x2]

    print(f'{number} x1,x2=({x1}, {x2}) border = {border}')
    return border


def is_part_number(number, lines):
    border = get_border(number, lines)

    border = filter(lambda x: not x.isdigit(), border)
    return not all(map(lambda x: x == '.', border))


print_lines(lines)

part_numbers = filter(lambda x: is_part_number(x, lines), numbers)
part_numbers = map(lambda x: x.value, part_numbers)
print(sum(part_numbers))


def has_border_on(number, lines, x, y):
    x1 = number.sidx-1
    x2 = number.sidx + len(number)

    if y == number.yidx - 1 and x1 <= x <= x2:
        return True
    if y == number.yidx + 1 and x1 <= x <= x2:
        return True

    if y == number.yidx and x == x1:
        return True
    if y == number.yidx and x == x2:
        return True

    return False


gears = []

for y in range(len(lines)):
    for x in range(len(lines[y])):
        if lines[y][x] == '*':
            neighbors = filter(lambda number: has_border_on(number, lines, x, y), numbers)
            neighbors = list(neighbors)

            if len(neighbors) == 2:
                print(f'Gear at {x}, {y} => {neighbors[0]} and {neighbors[1]}')
                gears.append((neighbors[0].value, neighbors[1].value))

print(sum(map(lambda x: x[0] * x[1], gears)))

File: 3/main.hs
#!/usr/bin/env stack
{- stack script
   --resolver lts-21.22
   --package split
   --package text
-}

import Data.Char (isDigit)
import Data.List (nub)
import Data.Maybe
import Debug.Trace (trace)
import System.IO.Unsafe (unsafePerformIO)
import Text.Read (readMaybe)

main :: IO ()
main = do
  input <- getContents
  let linesList = topAndBotton $ lines input

  -- mapM_ print linesList

  let grid = map cellsFromLine $ zip [0 ..] linesList
  let numbers = concatMap parseNumbersFromLine grid

  let p1Nums = map n $ filter (not . all (== '.') . numberNeighborsAsString grid) numbers
  -- print p1Nums
  print $ sum p1Nums

  return ()

topAndBotton :: [String] -> [String]
topAndBotton lines = map leftAndRight $ line : lines ++ [line]
  where
    line = replicate (length $ head lines) '.'

leftAndRight :: String -> String
leftAndRight line = "." ++ line ++ "."

data Coord = Coord {x :: Int, y :: Int} deriving (Show, Eq)

data Cell = Cell {coord :: Coord, char :: Char} deriving (Show, Eq)

data Number = Number {n :: Int, c :: Coord} deriving (Show, Eq)

parseNumbersFromLine :: [Cell] -> [Number]
parseNumbersFromLine cells = unsafePerformIO $ do numbers [] cells
  where
    numbers :: [Number] -> [Cell] -> IO [Number]
    numbers acc [] = return acc
    numbers acc (c : cs) = do
      if isDigit $ char c
        then do
          let toConvert = c : takeWhile (isDigit . char) cs
          let rest = dropWhile (isDigit . char) cs
          let number = cellToNumbers toConvert

          numbers (number : acc) rest
        else numbers acc cs

cellToNumbers :: [Cell] -> Number
cellToNumbers cells = foldl accumulate base cells
  where
    accumulate = (\acc cell -> Number {n = n acc * 10 + (read $ [char cell] :: Int), c = c acc})
    base = (Number {n = 0, c = coord $ head cells})

cellsFromLine :: (Int, String) -> [Cell]
cellsFromLine (yidx, line) = map (uncurry Cell) $ zip coords line
  where
    coords = coordsFromYidx (yidx, length line)

coordsFromYidx :: (Int, Int) -> [Coord]
coordsFromYidx (yidx, l) = map (`Coord` yidx) $ take l [0 ..]

numberNeighborsAsString :: [[Cell]] -> Number -> String
numberNeighborsAsString grid number = map lookup coords
  where
    coords = numberNeighbors number
    lookup :: Coord -> Char
    lookup coord = char $ grid !! y coord !! x coord

numberNeighbors :: Number -> [Coord]
numberNeighbors number = neighborhood coords
  where
    numLen = length $ show $ n number
    indexes = [x $ c number ..]
    xcoord = take numLen indexes
    coords = map (\x -> Coord {x = x, y = y $ c number}) xcoord

neighbors :: (Int, Int) -> [(Int, Int)]
neighbors (x, y) = filter (\t -> t /= (x, y)) $ do
  x' <- [x - 1, x, x + 1]
  y' <- [y - 1, y, y + 1]
  return (x', y')

coordNeighbors :: Coord -> [Coord]
coordNeighbors coord = map xyToCoords $ neighbors (x coord, y coord)

xyToCoords :: (Int, Int) -> Coord
xyToCoords (x, y) = Coord {x = x, y = y}

neighborhood :: [Coord] -> [Coord]
neighborhood coords = nub $ filter (`notElem` coords) $ concatMap coordNeighbors coords
